### ISSUE TITLE
Support for macOS 10.12

### ISSUE THREAD

**Issue by elmcapp** [chipjarred/MacMenuBar#8]
Just notice that current system requirement for this package is 10.15. Is there any way to get the package to work with 10.12

**Comment by chipjarred** [comment]
I can't think of any specific dependencies on macOS 10.15 features that aren't also available on 10.12 (or even earlier).  Pretty much all OS specific code is based on very old parts of Cocoa that go back to NextStep.   The important dependency is on the Swift version, because MacMenuBar needs result builders which were introduced in Swift 5.1 for its SwiftUI-like API.   Off the top of my head, I don't recall the reason I set 10.15 as the minimum macOS version.  It may have been something to do with Xcode's Swift version and its dependency on 10.15, or it may have just been that it seemed reasonable for a new package.  I'm certainly open to changing it, if it doesn't cause any problems.

Please leave this issue open, as it will remind me to take a closer look at it.  I plan to dedicate some time to updating MacMenuBar soon.   Presumably you're asking, because you need macOS 10.12 compatibility now.  I see that you forked the project, so you can try it in a branch of your fork.  Just change the macOS version in `Package.swift` to `.v10_12`, and of course set the package dependency in Xcode to use that branch. It should work, but it will be helpful when I do the update to know the result of your trying it, while allowing you to make progress on your project in the mean time.

**Comment by elmcapp** [comment]
To make it work with macOS version 10.12 there are 2 changes that would have to be made.

in action.swift
Some is only supported in macOS 10.15 and newer
````
public extension Action
{
    static var none: some Action { NoAction() }
    
    // -------------------------------------
    /** other code below
}
````
show be changed to
````
public extension Action
{
    static var none: some Action { NoAction() }
    
    // -------------------------------------
    /** other code below
}
````

Also you would have to use a different method for this code block as it's only supported in macOS 10.12.2 and newer
````
case .textCompletion:
                return (
                    #selector(NSTextView.toggleAutomaticTextCompletion(_:)),
                    .none
                )

````

**Comment by chipjarred** [comment]
Replying from my phone this time so I’m limited in looking at my source tree until I get back to my Mac later.  For the first change, I’d need to see if Action had `Self` or associated type constraints. If it doesn’t, `some` can be deleted, and `as Action` added to the return value.  That will be less efficient, but menu actions aren’t invoked that frequently for efficiency to matter that much. 

The thing is that SwiftUI uses `some`, and MacMenuBar is intended to be used in a SwiftUI app.  I can see how it might be useful in an old school Cocoa app though.  I might add some conditional compilation when I do my update. 

The second change will require more thought, but for now  if you don’t need automatic text completion for you could just `#if` that case in your fork or test for availability  and do nothing if it’s not available.   Those are just band-aids though.   I’d want something less special case-ish for a permanent solution.

**Comment by elmcapp** [comment]
what is toggleAutomaticTextCompletion used for. 

It would be nice if you could add some conditions in the update. I started working on a solution for React Native for macOS. It was brought up a few times that people would like to have control of menus within React native for macOS. I was planning to incorporate your package as a solution.

**Comment by chipjarred** [comment]
As you might guess, `toggleAutomaticTextCompletion` toggles whether macOS’s autocompletion feature is enabled or disabled for the receiving `NSTextView`.  In my opinion, It’s a less useful feature on macOS than on iOS for most text views. The line in question specifies that method’s selector as the menu item action handler.   Most of those standard action selectors are taken from the menu provided in Xcode’s Cocoa app project template’s storyboard.

**Comment by chipjarred** [comment]
I've pushed changes to allow usage back to macOS 10.12.   Let me know if that works for you.

**Comment by chipjarred** [comment]
@elmcapp, from my perspective I think I can close this issue since I made changes to allow MacMenuBar to be used going back to macOS 10.12.  However, since you are the one who reported it, before I close it, I'd like confirmation from you that it addresses the problem in your code base.

Something to be aware of when you're using MacMenuBar, in the time since I wrote it, SwiftUI has added its own `KeyEquivalent` type that results in ambiguity when trying to reference MacMenuBar's `KeyEquivalent`.  It would only affect you if you use `MacMenuBar.KeyEquivalent` for custom buttons or pop-up menus in a SwiftUI app.  You probably don't, but I wanted to mention it, since it's a SwiftUI change that broke the Soduko example application's build (which I just fixed).
